stellargraph/connector/neo4j/mapper.py

```diff
@@ -0,0 +1,251 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__all__ = [
+    "Neo4JGraphSAGENodeGenerator",
+    "Neo4JDirectedGraphSAGENodeGenerator",
+]
+
+import warnings
+import numpy as np
+import itertools as it
+
+from .sampler import (
+    Neo4JDirectedBreadthFirstNeighbors,
+    Neo4JSampledBreadthFirstWalk,
+)
+
+from ...core.graph import StellarGraph, GraphSchema
+from ...mapper import NodeSequence
+from ...mapper.sampled_node_generators import BatchedNodeGenerator
+from ...core.experimental import experimental
+
+
+@experimental(reason="the class is not fully tested")
+class Neo4JBatchedNodeGenerator(BatchedNodeGenerator):
+    """
+    Abstract base class for graph data generators from Neo4J.
+
+    The supplied graph should be a StellarGraph object that is ready for
+    machine learning. Currently the model requires node features for all
+    nodes in the graph.
+
+    Do not use this base class: use a subclass specific to the method.
+
+    Args:
+        G (StellarGraph): The machine-learning ready graph.
+        batch_size (int): Size of batch to return.
+        neo4j_graphdb (py2neo.Graph): the Neo4J Graph Database object
+        schema (GraphSchema): [Optional] Schema for the graph, for heterogeneous graphs.
+    """
+
+    def __init__(self, G, batch_size, neo4j_graphdb, schema=None):
+
+        import py2neo
```

Comment:
`py2neo` isn't used inside this `__init__`, so it doesn't _strictly_ need to be imported. Are you importing it so that the user gets an error about needing `py2neo` as soon as possible? If you are, it would be good to have a comment here, or even intercept the error to improve the message:

```python
# check the py2neo module is installed (if it isn't, this code is unlikely to work)
try:
    import py2neo
except ModuleNotFoundError as e:
    raise ModuleNotFoundError(f"{e.msg}. StellarGraph can only connect to Neo4j using the 'py2neo' module; please install it", name=e.name, path=e.path) from None
```

<details><summary>What's the `from None` doing? (click for details)</summary>

It's making sure the user only sees one `ModuleNotFoundError` by suppressing the "context" of the first exception being caught. See https://docs.python.org/3/library/exceptions.html

No `from None`:

```python
try:
    import py2neo
except ModuleNotFoundError as e:
    raise ModuleNotFoundError(f"{e.msg}. StellarGraph can only connect to Neo4j using the 'py2neo' module; please install it", name=e.name, path=e.path)
```
```
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-20-56fdd076cde9> in <module>
      1 try:
----> 2     import py2neo
      3 except ModuleNotFoundError as e:

ModuleNotFoundError: No module named 'py2neo'

During handling of the above exception, another exception occurred:

ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-20-56fdd076cde9> in <module>
      2     import py2neo
      3 except ModuleNotFoundError as e:
----> 4     raise ModuleNotFoundError(f"{e.msg}. StellarGraph can only connect to Neo4j using the 'py2neo' module; please install it", name=e.name, path=e.path)

ModuleNotFoundError: No module named 'py2neo'. StellarGraph can only connect to Neo4j using the 'py2neo' module; please install it
```

With `from None`:

```python
try:
    import py2neo
except ModuleNotFoundError as e:
    raise ModuleNotFoundError(f"{e.msg}. StellarGraph can only connect to Neo4j using the 'py2neo' module; please install it", name=e.name, path=e.path) from None
```

```
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-21-cbbd8be25d60> in <module>
      2     import py2neo
      3 except ModuleNotFoundError as e:
----> 4     raise ModuleNotFoundError(f"{e.msg}. StellarGraph can only connect to Neo4j using the 'py2neo' module; please install it", name=e.name, path=e.path) from None

ModuleNotFoundError: No module named 'py2neo'. StellarGraph can only connect to Neo4j using the 'py2neo' module; please install it
```

</details>


---

We should also consider checking the type of the `neo4j_graphdb` argument, e.g.:

```python
if not isinstance(neo4j_graphdb, py2neo.Graph):
    raise TypeError(f"neo4j_graphdb: expected py2neo.Graph, found {type(neo4j_graphdb)}")

Reply:
Yes. I'll add import exception handling and type-checking now!